File: ruby/sort/counting_sort.rb
def counting_sort(array)
  min = max = array[0]
  (1...array.size).each do |x|
    min = x if x < min
    max = x if x > max
  end

  arr = Array.new(max - min + 1, 0)
  array.each do |x|
    arr[x - min] += 1
  end

  index = 0
  arr.each_with_index do |c, i|
    c.times do
      array[index] = min + i
      index += 1
    end
  end

  array
end

# array = 100.times.map { rand(100) }
# array1 = array.sort
# array2 = counting_sort(array)

# p array1
# p array2

# fail unless array1 == array2

File: javascript/sort/countingSort.js
function countingSort(array) {
  let min
  let max
  min = max = array[0]

  for (const x of array) {
    if (x < min) min = x
    if (x > max) max = x
  }

  arr = new Array(max - min + 1)
  for (const x of array) {
    arr[x - min] = (arr[x - min] || 0) + 1
  }

  let index = 0
  for (let i = 0; i < arr.length; ++i) {
    const c = arr[i]
    if (!c) continue
    for (let j = 0; j < c; ++j) {
      array[index] = min + i
      index += 1
    }
  }

  return array
}

// let arr = [...new Array(20)].map(_ => Math.round(Math.random() * 99))
// console.log(countingSort(arr))
